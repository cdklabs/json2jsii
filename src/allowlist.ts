export const NAMED_SYMBOLS = {
  EXCLAMATION: '!',
  AT_SIGN: '@',
  BACKTICK: '`',
  AMPERSAND: '&',
  APOSTROPHE: '\'',
  ASTERISK: '*',
  DOLLAR_SIGN: '$',
  POUND_SIGN: '#',
  TILDE: '~',
  PERCENT: '%',
  CARAT: '^',
  PIPE: '|',
  QUESTION_MARK: '?',
  COMMA: ',',
  PERIOD: '.',
  LESS_THAN: '<',
  GREATER_THAN: '>',
  OPEN_BRACKET: '(',
  CLOSE_BRACKET: ')',
  OPEN_BRACE: '{',
  CLOSE_BRACE: '}',
  SEMI_COLON: ';',
  QUOTATION_MARK: '"',
  FORWARD_SLASH: '/',
  BACKWARD_SLASH: '\\',
  NOT_EQUALS_TO: '!=',
  EQUALS_TO: '==',
  DEEP_EQUALS: '===',
  EQUAL_TILDE: '=~',
  NEGATION_TILDE: '!~',
  ROUND_BRACKETS: '()',
  CURLY_BRACES: '{}',
  ANGLE_BRACKETS: '<>',
  GREATER_THAN_EQUAL_TO: '>=',
  LESSER_THAN_EQUAL_TO: '<=',
};