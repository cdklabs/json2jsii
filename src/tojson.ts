import { Code } from './code';

export class ToJsonFunction {
  /**
   * The name the toJson function for a struct.
   */
  public readonly functionName: string;

  private readonly fields: Record<string, string> = {};

  constructor(private readonly baseType: string, private readonly internal: boolean = false) {
    this.functionName = `toJson_${baseType}`;
  }

  /**
   * Adds a field to the struct.
   *
   * @param schemaName The name of the property in the schema ("to")
   * @param propertyName The name of the TypeScript property ("from")
   * @param toJson A function used to convert a value from JavaScript to schema
   * format. This could be `x => x` if no conversion is required.
   */
  public addField(schemaName: string, propertyName: string, toJson: ToJson) {
    this.fields[schemaName] = toJson(`obj.${propertyName}`);
  }

  public emit(code: Code) {
    const disabledEslintRules = [
      'max-len',
      '@stylistic/max-len',
      'quote-props',
      '@stylistic/quote-props',
    ];

    code.line();
    code.line('/**');
    code.line(` * Converts an object of type '${this.baseType}' to JSON representation.`);
    if (this.internal) {
      code.line(' * @internal');
    }
    code.line(' */');
    code.line(`/* eslint-disable ${disabledEslintRules.join(', ')} */`);
    code.openBlock(`export function ${this.functionName}(obj: ${this.baseType} | undefined): Record<string, any> | undefined`);
    code.line('if (obj === undefined) { return undefined; }');

    code.open('const result = {');
    for (const [k, v] of Object.entries(this.fields)) {
      code.line(`'${k}': ${v},`);
    }
    code.close('};');

    code.line('// filter undefined values');
    code.line('return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});');

    code.closeBlock();
    code.line(`/* eslint-enable ${disabledEslintRules.join(', ')} */`);
  }
}

/**
 * A function that converts an expression from JavaScript to schema format.
 *
 * @example x => x
 * @example x => x?.map(y => toJson_Foo(y))
 */
export type ToJson = (expression: string) => string;
