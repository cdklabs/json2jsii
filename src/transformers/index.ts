export * from './null-in-type-array';
export * from './null-in-union';
export * from './union-duplicates';
export * from './singleton-unions';
export * from './array-union-reducer';
