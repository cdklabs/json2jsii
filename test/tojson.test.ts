
import { spawnSync } from 'child_process';
import { mkdtempSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { JSONSchema4 } from 'json-schema';
import { TypeGenerator } from '../src';

test('primitives', () => {
  const toJson = generateToJson({
    properties: {
      StringProperty: { type: 'string' },
      BooleanProperty: { type: 'boolean' },
      NumberProperty: { type: 'number' },
      With_UnderScore: { type: 'integer' },
    },
    required: [
      'StringProperty',
    ],
  });

  expect(toJson({
    stringProperty: 'hello', booleanProperty: false, numberProperty: 1234, withUnderScore: 8989,
  })).toStrictEqual({
    StringProperty: 'hello', BooleanProperty: false, NumberProperty: 1234, With_UnderScore: 8989,
  });

  expect(toJson({
    stringProperty: 'hello, world', numberProperty: undefined,
  })).toStrictEqual({
    StringProperty: 'hello, world',
  });
});

test('names can get crazy', () => {
  const toJson = generateToJson({
    properties: {
      'hyphen-name': { type: 'string' },
      '$nameWithDolar': { type: 'string' },
      'name with spaces': { type: 'string' },
      'x-extension': { type: 'string' },
    },
  });

  expect(toJson({
    hyphenName: 'boomboom',
    nameWithDolar: 'hey',
    nameWithSpaces: 'glow',
    xExtension: 'booom',
  })).toStrictEqual({
    'hyphen-name': 'boomboom',
    '$nameWithDolar': 'hey',
    'name with spaces': 'glow',
    'x-extension': 'booom',
  });
});

test('complex types', () => {
  const toJson = generateToJson({
    definitions: {
      YourType: {
        properties: {
          John: { type: 'array', items: { type: 'string' } },
        },
      },
      MyType: {
        properties: {
          Foo: { type: 'number' },
          Bar: { type: 'string' },
          Nested: { $ref: '#/definitions/YourType' },
        },
        required: ['Nested'],
      },
    },
    properties: {
      ComplexType: { $ref: '#/definitions/MyType' },
    },
  });

  expect(toJson({
    complexType: { foo: 1234, bar: 'hey there', nested: { john: ['foo'] } },
  })).toStrictEqual({
    ComplexType: { Foo: 1234, Bar: 'hey there', Nested: { John: ['foo'] } },
  });
});

describe('arrays', () => {

  test('of primitives', () => {
    const toJson = generateToJson({
      properties: {
        ArrayProp: { type: 'array', items: { type: 'string' } },
      },
    });

    expect(toJson({
      arrayProp: ['bello', 'aoll'],
    })).toStrictEqual({
      ArrayProp: ['bello', 'aoll'],
    });
  });

  test('of complex types', () => {
    const toJson = generateToJson({
      definitions: {
        MyType: {
          properties: {
            Foo: { type: 'number' },
            Bar: { type: 'string' },
          },
        },
      },
      properties: {
        Array_Of_complex: { type: 'array', items: { $ref: '#/definitions/MyType' } },
      },
    });

    expect(toJson({
      arrayOfComplex: [{ foo: 122 }, { bar: 'hello' }, { foo: 88, bar: 'world' }],
    })).toStrictEqual({
      Array_Of_complex: [{ Foo: 122 }, { Bar: 'hello' }, { Foo: 88, Bar: 'world' }],
    });
  });

  test('without items', () => {
    const toJson = generateToJson({
      properties: {
        ArrayProp: { type: 'array' },
      },
    });

    expect(toJson({
      arrayProp: ['bello', { foo: 88, bar: 'world' }],
    })).toStrictEqual({
      ArrayProp: ['bello', { foo: 88, bar: 'world' }],
    });
  });

  test('with items', () => {
    const toJson = generateToJson({
      properties: {
        ArrayProp: {
          type: 'array',
          items: {
            type: 'string',
          },
        },
      },
    });

    expect(toJson({
      arrayProp: ['bello', { foo: 88, bar: 'world' }],
    })).toStrictEqual({
      ArrayProp: ['bello', { foo: 88, bar: 'world' }],
    });
  });

});

test('any', () => {
  const toJson = generateToJson({
    properties: {
      MyAny: { type: 'object' },
    },
  });

  expect(toJson({ myAny: 177 })).toStrictEqual({ MyAny: 177 });
  expect(toJson({ myAny: 'hello' })).toStrictEqual({ MyAny: 'hello' });
  expect(toJson({ myAny: ['hello', 889] })).toStrictEqual({ MyAny: ['hello', 889] });
});


describe('maps', () => {
  test('of primitives', () => {
    const toJson = generateToJson({
      properties: {
        StringMap: { additionalProperties: { type: 'string' } },
        NumberMap: { additionalProperties: { type: 'number' } },
      },
    });

    expect(toJson({
      stringMap: {
        hello: 'string',
        world: 'strong',
      },
      numberMap: {
        hello: 10,
        world: 20,
      },
    })).toStrictEqual({
      StringMap: {
        hello: 'string',
        world: 'strong',
      },
      NumberMap: {
        hello: 10,
        world: 20,
      },
    });

    expect(toJson({
      stringMap: undefined,
      numberMap: {
        hello: undefined,
        world: 11,
      },
    })).toStrictEqual({
      NumberMap: {
        world: 11,
      },
    });
  });

  test('of complex', () => {
    const toJson = generateToJson({
      definitions: {
        MyType: {
          properties: {
            Foo: { type: 'number' },
            Bar: { type: 'string' },
          },
        },
      },
      properties: {
        ComplexMap: { additionalProperties: { $ref: '#/definitions/MyType' } },
      },
    });

    expect(toJson({
      complexMap: {
        foya: { foo: 123, bar: 'barbar' },
        hello_world: { foo: 3333 },
      },
    })).toStrictEqual({
      ComplexMap: {
        foya: { Foo: 123, Bar: 'barbar' },
        hello_world: { Foo: 3333 },
      },
    });
  });
});

test('enums', () => {
  const toJson = generateToJson({
    properties: {
      MyEnum: { enum: ['one', 'two', 'three'] },
      YourEnum: { enum: ['jo', 'shmo'] },
    },
  });

  expect(toJson({
    myEnum: 'two',
    yourEnum: undefined,
  })).toStrictEqual({
    MyEnum: 'two',
  });
});

test('date', () => {
  const toJson = generateToJson({
    properties: {
      DateTime: { type: 'string', format: 'date-time' },
    },
  });

  expect(toJson({
    dateTime: new Date('2021-07-19T12:08:52.210Z'),
  })).toStrictEqual({
    DateTime: '2021-07-19T12:08:52.210Z',
  });

  expect(toJson({ dateTime: undefined })).toStrictEqual({});
});

test('union types', () => {
  const module = generateModule('MyStruct', {
    definitions: {
      ComplexType: {
        properties: {
          Ref_to_union: { $ref: '#/definitions/IntOrString' },
        },
      },
      IntOrString: {
        oneOf: [
          { type: 'integer' },
          { type: 'string' },
        ],
      },
    },
    properties: {
      ref_to_complex: { $ref: '#/definitions/ComplexType' },
      ReusedType: { $ref: '#/definitions/IntOrString' },
      Haver: {
        anyOf: [
          { type: 'integer' },
          { type: 'string' },
        ],
      },
    },
  });

  const toJson = module.toJson_MyStruct;
  const MyStructHaver = module.MyStructHaver;
  const IntOrString = module.IntOrString;

  expect(toJson({
    haver: MyStructHaver.fromString('hello'),
    refToComplex: {
      refToUnion: IntOrString.fromNumber(1234),
    },
  })).toStrictEqual({
    Haver: 'hello',
    ref_to_complex: {
      Ref_to_union: 1234,
    },
  });
});

/**
 * Generates structs from this JSON schema and returns a compiled versio of the
 * toJson method for the top-level struct.
 */
function generateToJson(schema: JSONSchema4): (x: any) => any {
  return generateModule('MyStruct', schema).toJson_MyStruct;
}

/**
 * Generates structs from this JSON schema and returns a compiled versio of the
 * toJson method for the top-level struct.
 */
function generateModule(structName: string, schema: JSONSchema4): any {
  // generate code into workdir
  const workdir = mkdtempSync(join(tmpdir(), 'tojson.'));

  const gen = TypeGenerator.forStruct(structName, schema);
  const code = gen.render();
  expect(code).toMatchSnapshot();

  writeFileSync(join(workdir, 'index.ts'), code);

  try {
    // compile using the typescript compiler
    const tsc = require.resolve('typescript/bin/tsc');
    spawnSync(tsc, ['index.ts'], { cwd: workdir });

    // import the compiled javascript code into this process (wow)
    const source = join(workdir, 'index.js');


    // eslint-disable-next-line @typescript-eslint/no-require-imports
    return require(source);
  } catch (e) {
    throw new Error(`Compilation error: ${e}. Workdir: ${workdir}`);
  }
}