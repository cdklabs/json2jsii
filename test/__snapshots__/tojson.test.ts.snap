// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`any 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#MyAny
   */
  readonly myAny?: any;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'MyAny': obj.myAny,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`arrays of complex types 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#Array_Of_complex
   */
  readonly arrayOfComplex?: MyType[];

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Array_Of_complex': obj.arrayOfComplex?.map(y => toJson_MyType(y)),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema MyType
 */
export interface MyType {
  /**
   * @schema MyType#Foo
   */
  readonly foo?: number;

  /**
   * @schema MyType#Bar
   */
  readonly bar?: string;

}

/**
 * Converts an object of type 'MyType' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyType(obj: MyType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Foo': obj.foo,
    'Bar': obj.bar,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`arrays of primitives 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ArrayProp
   */
  readonly arrayProp?: string[];

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ArrayProp': obj.arrayProp?.map(y => y),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`complex types 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ComplexType
   */
  readonly complexType?: MyType;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ComplexType': toJson_MyType(obj.complexType),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema MyType
 */
export interface MyType {
  /**
   * @schema MyType#Foo
   */
  readonly foo?: number;

  /**
   * @schema MyType#Bar
   */
  readonly bar?: string;

  /**
   * @schema MyType#Nested
   */
  readonly nested: YourType;

}

/**
 * Converts an object of type 'MyType' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyType(obj: MyType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Foo': obj.foo,
    'Bar': obj.bar,
    'Nested': toJson_YourType(obj.nested),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema YourType
 */
export interface YourType {
  /**
   * @schema YourType#John
   */
  readonly john?: string[];

}

/**
 * Converts an object of type 'YourType' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_YourType(obj: YourType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'John': obj.john?.map(y => y),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`date 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#DateTime
   */
  readonly dateTime?: Date;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'DateTime': obj.dateTime?.toISOString(),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`enums 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#MyEnum
   */
  readonly myEnum?: MyStructMyEnum;

  /**
   * @schema MyStruct#YourEnum
   */
  readonly yourEnum?: MyStructYourEnum;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'MyEnum': obj.myEnum,
    'YourEnum': obj.yourEnum,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema MyStructMyEnum
 */
export enum MyStructMyEnum {
  /** one */
  ONE = 'one',
  /** two */
  TWO = 'two',
  /** three */
  THREE = 'three',
}

/**
 * @schema MyStructYourEnum
 */
export enum MyStructYourEnum {
  /** jo */
  JO = 'jo',
  /** shmo */
  SHMO = 'shmo',
}
"
`;

exports[`maps of complex 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ComplexMap
   */
  readonly complexMap?: { [key: string]: MyType };

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ComplexMap': ((obj.complexMap) === undefined) ? undefined : (Object.entries(obj.complexMap).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: toJson_MyType(i[1]) }), {})),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema MyType
 */
export interface MyType {
  /**
   * @schema MyType#Foo
   */
  readonly foo?: number;

  /**
   * @schema MyType#Bar
   */
  readonly bar?: string;

}

/**
 * Converts an object of type 'MyType' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyType(obj: MyType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Foo': obj.foo,
    'Bar': obj.bar,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`maps of primitives 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#StringMap
   */
  readonly stringMap?: { [key: string]: string };

  /**
   * @schema MyStruct#NumberMap
   */
  readonly numberMap?: { [key: string]: number };

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'StringMap': ((obj.stringMap) === undefined) ? undefined : (Object.entries(obj.stringMap).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {})),
    'NumberMap': ((obj.numberMap) === undefined) ? undefined : (Object.entries(obj.numberMap).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {})),
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`names can get crazy 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#hyphen-name
   */
  readonly hyphenName?: string;

  /**
   * @schema MyStruct#$nameWithDolar
   */
  readonly nameWithDolar?: string;

  /**
   * @schema MyStruct#name with spaces
   */
  readonly nameWithSpaces?: string;

  /**
   * @schema MyStruct#x-extension
   */
  readonly xExtension?: string;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'hyphen-name': obj.hyphenName,
    '$nameWithDolar': obj.nameWithDolar,
    'name with spaces': obj.nameWithSpaces,
    'x-extension': obj.xExtension,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`primitives 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#StringProperty
   */
  readonly stringProperty: string;

  /**
   * @schema MyStruct#BooleanProperty
   */
  readonly booleanProperty?: boolean;

  /**
   * @schema MyStruct#NumberProperty
   */
  readonly numberProperty?: number;

  /**
   * @schema MyStruct#With_UnderScore
   */
  readonly withUnderScore?: number;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'StringProperty': obj.stringProperty,
    'BooleanProperty': obj.booleanProperty,
    'NumberProperty': obj.numberProperty,
    'With_UnderScore': obj.withUnderScore,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */
"
`;

exports[`union types 1`] = `
"/**
 * @schema MyStruct
 */
export interface MyStruct {
  /**
   * @schema MyStruct#ref_to_complex
   */
  readonly refToComplex?: ComplexType;

  /**
   * @schema MyStruct#ReusedType
   */
  readonly reusedType?: IntOrString;

  /**
   * @schema MyStruct#Haver
   */
  readonly haver?: MyStructHaver;

}

/**
 * Converts an object of type 'MyStruct' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_MyStruct(obj: MyStruct | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'ref_to_complex': toJson_ComplexType(obj.refToComplex),
    'ReusedType': obj.reusedType?.value,
    'Haver': obj.haver?.value,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema ComplexType
 */
export interface ComplexType {
  /**
   * @schema ComplexType#Ref_to_union
   */
  readonly refToUnion?: IntOrString;

}

/**
 * Converts an object of type 'ComplexType' to JSON representation.
 */
/* eslint-disable max-len, quote-props */
export function toJson_ComplexType(obj: ComplexType | undefined): Record<string, any> | undefined {
  if (obj === undefined) { return undefined; }
  const result = {
    'Ref_to_union': obj.refToUnion?.value,
  };
  // filter undefined values
  return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});
}
/* eslint-enable max-len, quote-props */

/**
 * @schema IntOrString
 */
export class IntOrString {
  public static fromNumber(value: number): IntOrString {
    return new IntOrString(value);
  }
  public static fromString(value: string): IntOrString {
    return new IntOrString(value);
  }
  private constructor(public readonly value: number | string) {
  }
}

/**
 * @schema MyStructHaver
 */
export class MyStructHaver {
  public static fromNumber(value: number): MyStructHaver {
    return new MyStructHaver(value);
  }
  public static fromString(value: string): MyStructHaver {
    return new MyStructHaver(value);
  }
  private constructor(public readonly value: number | string) {
  }
}
"
`;
